bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -0,0 +1,171 @@
+package com.bugsnag.android.internal
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import java.io.File
+import java.io.IOException
+import java.lang.IllegalStateException
+import java.nio.BufferOverflowException
+import java.util.function.BiConsumer
+
+/**
+ * A document with journal backing.
+ * Document snapshots are persisted to files, and journal entries are persisted to a memory-mapped
+ * file.
+ *
+ * Changes to the document must me made only via journal commands. Do not modify the document or any
+ * of its sub-components directly!
+ */
+class JournaledDocument
+/**
+ * Constructor.
+ * Upon construction, this document will have associated snapshot and journal files.
+ *
+ * @param baseDocumentPath The base path + filename to derive snapshot and journal names from.
+ * @param journalType A string identifying the type of journal (used for verification when the
+ *                    journal is reloaded)
+ * @param version The journal type version (to support migration)
+ * @param bufferSize The size of the shared memory buffer (and thus the maximum size of the journal)
+ * @param initialDocument The initial document contents (will be shallow copied)
+ */
+constructor(
+    baseDocumentPath: File,
+    journalType: String,
+    version: Int,
+    bufferSize: Long,
+    initialDocument: MutableMap<String, Any>
+) : Map<String, Any> {
```

Comment:
Is this inheritance hierarchy necessary? Methods from `Map` are unused in the test code and it seems like it'd be preferable to use composition rather than inheritance as this would hide the implementation detail.

Reply:
It's an interface implementation. I want the document to be usable like a map (i.e. `myDocument["somekey"]`)

Reply:
The map interface is checked here: https://github.com/bugsnag/bugsnag-android/pull/1314/files#diff-b1235e74d61e2d17abb375db5d7c87ffc872a3908e1a01d2a7d1ea12fdb20f50R84

Reply:
It might be worth also implementing `Closable` for clarity - and for the Kotlin `use` function